lib/runner.js

```diff
@@ -910,6 +917,11 @@ Runner.prototype.run = function(fn) {
   this.on(constants.EVENT_RUN_END, function() {
     debug(constants.EVENT_RUN_END);
     process.removeListener('uncaughtException', uncaught);
+    process.on('uncaughtException', function(err) {
+      // TODO: remove this listener when Node v8 support is dropped
+      process.exitCode = 1;
+      console.error('\n', err);
```

Comment:
I believe it should be `throw err`, as otherwise process will not be stopped

Reply:
You are right about the process. With `throw err` the process stops as expected, but with `exitCode: 7`. I haven't succeeded in overwriting this code, even `process.exitCode` doesn't help.
I decided therefore to leave with `process.exit(1)` to copy Node's default behavior as close as possible.

Reply:
Node calls my exit strategy :

> a misuse of the process.exit() method that could lead to data printed to stdout being truncated and lost:
[...]
The reason this is problematic is because writes to process.stdout in Node.js are sometimes asynchronous and may occur over multiple ticks of the Node.js event loop. Calling process.exit(), however, forces the process to exit before those additional writes to stdout can be performed.

Maybe I should really use `throw err`...

Reply:
Interesting, but I'm not sure if `throw err` makes real difference.

Also it's not clear in what cases it'll be an async (how to reproduce the issue?) I've just tried following on macOS

```javascript
process.stdout.write("foo".repeat(2 ** 27) + "\nEND\n");
process.exit(1);
```

And always I get all output written, maybe it depends on a system.

Reply:
It depends on [here](https://nodejs.org/api/process.html#process_process_stdout).
We have several output truncation issues in Mocha.

Reply:
It says that pipes are async on POSIX, not sure how reliable is below example, but i truncates same way with `process.exit` and `throw`

```javascript
const { Duplex } = require("stream");

class MyDuplex extends Duplex {
    _write(chunk, encoding, callback) {
        this.written = true;
        this.push(chunk);
        callback();
    }

    _read(size) {
        if (this.written) this.push(null);
    }
}

const interimStream = new MyDuplex();
interimStream.pipe(process.stdout);
interimStream.write("foo".repeat(2 ** 17) + "\nEND\n");
// Uncomment to prevent any output
// process.exit(1);
```